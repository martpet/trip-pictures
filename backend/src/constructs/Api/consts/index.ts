export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
