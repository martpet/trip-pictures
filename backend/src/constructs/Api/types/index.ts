export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
