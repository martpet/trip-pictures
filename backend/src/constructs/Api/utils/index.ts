export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
