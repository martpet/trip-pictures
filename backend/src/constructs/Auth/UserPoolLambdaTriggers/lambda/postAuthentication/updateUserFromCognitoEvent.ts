import {
  createDynamoUpdateExpression,
  DynamoDBClient,
  DynamoDBDocumentClient,
  filterChangedProps,
  GetCommand,
  PostAuthenticationTriggerEvent,
  UpdateCommand,
  UserPropsFromCognito,
  UsersTableItem,
  usersTableOptions,
} from 'lambda-layer';

import { getUserPropsFromCognitoEvent } from '../getUserPropsFromCognitoEvent';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient, {
  marshallOptions: { removeUndefinedValues: true },
});

export const updateUserFromCognitoEvent = async (
  event: PostAuthenticationTriggerEvent
) => {
  const propsFromEvent = getUserPropsFromCognitoEvent(event);
  const user = await fetchUser(propsFromEvent);

  if (!user) {
    throw new Error(`could not find user with id "${propsFromEvent.id}"`);
  }

  const changedProps = filterChangedProps(propsFromEvent, user);

  if (!changedProps) {
    return undefined;
  }

  const updateCommand = new UpdateCommand({
    ...makeTableParams(user),
    ...createDynamoUpdateExpression(changedProps),
  });

  return ddbDocClient.send(updateCommand);
};

async function fetchUser(props: UserPropsFromCognito) {
  const getCommand = new GetCommand(makeTableParams(props));
  const { Item } = await ddbDocClient.send(getCommand);
  return Item;
}

function makeTableParams(props: Partial<UsersTableItem>) {
  const pkName = usersTableOptions.partitionKey.name;
  return {
    TableName: usersTableOptions.tableName,
    Key: { [pkName]: props[pkName] },
  };
}
