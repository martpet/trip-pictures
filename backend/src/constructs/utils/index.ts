export * from './createEdgeFunction';
export * from './createNodejsFunction';
export * from './createTable';
