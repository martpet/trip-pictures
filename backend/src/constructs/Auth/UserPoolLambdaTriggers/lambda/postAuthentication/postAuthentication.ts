import { PostAuthenticationTriggerHandler } from 'lambda-layer';

import { updateUserFromCognitoEvent } from './updateUserFromCognitoEvent';

export const handler: PostAuthenticationTriggerHandler = async (event) => {
  await updateUserFromCognitoEvent(event);
  return event;
};
