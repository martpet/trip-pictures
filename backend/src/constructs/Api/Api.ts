import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
