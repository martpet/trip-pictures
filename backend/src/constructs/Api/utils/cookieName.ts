import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
