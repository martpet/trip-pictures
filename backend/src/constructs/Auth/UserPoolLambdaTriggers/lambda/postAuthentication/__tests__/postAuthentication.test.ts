import { itCalls, itResolves } from '~/constructs/Api/utils';

import { handler } from '../postAuthentication';
import { updateUserFromCognitoEvent } from '../updateUserFromCognitoEvent';
import event from './__fixtures__/postAuthenticationEvent';

vi.mock('../updateUserFromCognitoEvent');

const args = [event] as unknown as Parameters<typeof handler>;

describe('postAuthentication', () => {
  itCalls(updateUserFromCognitoEvent, handler, args);
  itResolves(handler, args);
});
