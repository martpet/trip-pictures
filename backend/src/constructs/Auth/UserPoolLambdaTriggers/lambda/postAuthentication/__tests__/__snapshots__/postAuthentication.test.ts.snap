// Vitest Snapshot v1

exports[`postAuthentication > calls "updateUserFromCognitoEvent" with correct args 1`] = `
[
  [
    {
      "callerContext": {
        "awsSdkVersion": "dummyPostAuthAwsSdkVersion",
        "clientId": "dummyPostAuthClientId",
      },
      "region": "dummyPostAuthRegion",
      "request": {
        "newDeviceUsed": false,
        "userAttributes": {
          "cognito:user_status": "EXTERNAL_PROVIDER",
          "email": "dummyPostAuthEmail",
          "email_verified": "false",
          "family_name": "dummyPostAuthFamilyName",
          "given_name": "dummyPostAuthGivenName",
          "identities": "[{\\"userId\\":\\"182347123047012\\",\\"providerName\\":\\"Google\\",\\"providerType\\":\\"Google\\",\\"issuer\\":null,\\"primary\\":true,\\"dateCreated\\":1658487089823}]",
          "picture": "dummyPostAuthPictureUrl",
          "sub": "dummyPostAuthSub",
        },
      },
      "response": {},
      "triggerSource": "PostAuthentication_Authentication",
      "userName": "dummyPostAuthUserName",
      "userPoolId": "dummyPostAuthUserPoolId",
      "version": "dummyPostAuthVersion",
    },
  ],
]
`;

exports[`postAuthentication > resolves with a correct value 1`] = `
{
  "callerContext": {
    "awsSdkVersion": "dummyPostAuthAwsSdkVersion",
    "clientId": "dummyPostAuthClientId",
  },
  "region": "dummyPostAuthRegion",
  "request": {
    "newDeviceUsed": false,
    "userAttributes": {
      "cognito:user_status": "EXTERNAL_PROVIDER",
      "email": "dummyPostAuthEmail",
      "email_verified": "false",
      "family_name": "dummyPostAuthFamilyName",
      "given_name": "dummyPostAuthGivenName",
      "identities": "[{\\"userId\\":\\"182347123047012\\",\\"providerName\\":\\"Google\\",\\"providerType\\":\\"Google\\",\\"issuer\\":null,\\"primary\\":true,\\"dateCreated\\":1658487089823}]",
      "picture": "dummyPostAuthPictureUrl",
      "sub": "dummyPostAuthSub",
    },
  },
  "response": {},
  "triggerSource": "PostAuthentication_Authentication",
  "userName": "dummyPostAuthUserName",
  "userPoolId": "dummyPostAuthUserPoolId",
  "version": "dummyPostAuthVersion",
}
`;
